### PR TITLE
AF: Extended training via reduced eval frequency (chihiro)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The AF champion (PR #3135, W&B run: sh2zyfwr) was still actively improving at timeout:
- Best surface at epoch 704 out of 763 total epochs
- Best volume at epoch 743 out of 763 total epochs
- Both metrics still on a downward trajectory at cutoff

AirfRANS evaluation is expensive — computing surface and volume metrics over the full dataset takes significant time per epoch. By reducing eval frequency from every epoch to every 3 epochs, we save ~67% of eval compute. This translates directly to MORE training epochs within the same wall-clock budget.

If the champion can train to 900–1000+ epochs (vs 763), the continued descent could naturally close the volume gap (0.002039 → target 0.0017, a 17% improvement) without any model or loss changes. This is the simplest possible approach to volume closure.

## Baseline (PR #3135, run sh2zyfwr)

| Metric | Value | Epoch |
|--------|-------|-------|
| val_primary/surface_mse | 0.000296 | 704 |
| vol_mse | 0.002039 | 743 |
| Total epochs | 763 (timeout) | — |

Config: 2L/256d/4H, AdamW lr=6e-4, T_max=50, gc=1.0, WD=1e-2, EMA=0.999, vol-weight=10x, Fourier

**Volume target: vol_mse < 0.0017 (SpiderSolver benchmark). Gap: 17%.**

## Known Bug Fix

There is a `primary_metric_key` shadowing bug in train.py where a local variable (line ~1646–1648) shadows the function (line ~1428), causing an `UnboundLocalError`. Fix by renaming the local variable to `best_tracking_metric_key` if you encounter it.

## Implementation

Add an `--eval-every-n-epochs` flag (integer, default 1):
1. Only run validation every N epochs. Skip eval on non-eval epochs.
2. Best-checkpoint saving still works — just updates less frequently.
3. Always eval on the LAST epoch before timeout (to capture terminal metrics).
4. Log `eval_every_n_epochs` to W&B config.

### Trial 1 — Eval every 3 epochs (champion config)

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --ema-decay 0.999 \
  --vol-loss-weight 10.0 \
  --eval-every-n-epochs 3 \
  --wandb_group af-extended-training
```

### Trial 2 — Eval every 5 epochs (more aggressive savings)

Same as Trial 1 but with `--eval-every-n-epochs 5`

## Results to Report

- Total epochs trained and wall-clock time for each trial
- Best val surface_mse and epoch for each trial
- Best val vol_mse and epoch for each trial
- Direct comparison to baseline sh2zyfwr (763 epochs, surface=0.000296 at ep704, vol=0.002039 at ep743)

Target: **vol_mse < 0.0017** while maintaining **surface_mse < 0.000350**.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT chihiro:

## Results

### Trial 1 — eval-every-3 (W&B: gtwobpiy)

| Metric | Baseline (sh2zyfwr) | Trial 1 | Δ |
|--------|---------------------|---------|---|
| val_primary/surface_mse | 0.000296 (ep704) | **0.000266** (ep792) | **−10.1%** ✅ |
| vol_mse | 0.002039 (ep743) | 0.002193 (ep792) | +7.6% |
| Total epochs | 763 (timeout) | **794** (timeout) | **+4.1%** |
| best_test/surface_mse | — | 0.000451 | — |
| best_test/volume_mse | — | 0.003223 | — |
| final_test/surface_mse | — | 0.000535 | — |
| final_test/volume_mse | — | 0.003033 | — |

**Command:**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=999 SENPAI_TIMEOUT_MINUTES=360.0 CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 --volume-loss-weight 10.0 \
  --eval-every-n-epochs 3 \
  --wandb-group af-extended-training --agent chihiro --wandb-name "chihiro/af-extended-eval3"
```

### Trial 2 — eval-every-5 (W&B: p4tkm502)

| Metric | Baseline (sh2zyfwr) | Trial 2 | Δ |
|--------|---------------------|---------|---|
| val_primary/surface_mse | 0.000296 (ep704) | **0.000288** (ep615) | **−2.7%** ✅ |
| vol_mse | 0.002039 (ep743) | 0.002897 (ep735) | +42.1% |
| Total epochs | 763 (timeout) | **790** (timeout) | **+3.5%** |
| best_test/surface_mse | — | 0.001933 | — |
| best_test/volume_mse | — | 0.004090 | — |

⚠️ **Trial 2 diverged at ~epoch 745.** Metrics were excellent through epoch 740 (srf=0.000348, vol=0.002969), then catastrophically degraded (srf>0.2 by epoch 760). The best checkpoint at epoch 615 is preserved but the model never recovered. This may be related to cosine restart instability at the 750 boundary with eval-every-5 potentially missing critical early-warning signals.

**Command:**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=999 SENPAI_TIMEOUT_MINUTES=360.0 CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 --volume-loss-weight 10.0 \
  --eval-every-n-epochs 5 \
  --wandb-group af-extended-training --agent chihiro --wandb-name "chihiro/af-extended-eval5"
```

### What happened

**Trial 1 (eval-every-3) is the clear winner.** It trained 794 epochs (vs baseline's 763) within the same 360-minute wall-clock budget, achieving a **10.1% improvement in surface_mse** (0.000266 vs 0.000296). The surface metric was still on a downward trajectory at timeout — the best was at epoch 792, very close to the end.

The volume metric (0.002193) did not beat the baseline's 0.002039, missing the target of 0.0017 by a significant margin. However, volume_mse was still on a clear downward trajectory through the run (from 0.013 at ep150 → 0.006 at ep300 → 0.003 at ep500 → 0.002 at ep790). More training time would very likely close this gap.

**Trial 2 (eval-every-5)** also beat the baseline on surface (0.000288 at ep615, −2.7%), and was on a strong trajectory (0.000348 at ep740), but it diverged catastrophically around epoch 745. The divergence was sudden — metrics went from 0.000348 to 0.005 in 5 epochs, then to 0.26 by epoch 760. The EMA model failed to recover. Root cause is likely numerical instability at high epoch counts, possibly exacerbated by less frequent checkpoint tracking with eval-every-5.

### Key findings

1. **Reduced eval frequency works.** eval-every-3 achieved 794 epochs vs 763 baseline — a 4% increase in training epochs within the same wall-clock budget.
2. **Surface metric improved significantly.** Trial 1 achieved 0.000266 (−10.1% vs baseline). The metric was still descending at timeout.
3. **Volume target not met.** Best vol_mse was 0.002193 (Trial 1 at ep792), vs target of 0.0017. The gap is 29%, but the trajectory suggests more training time could close it.
4. **eval-every-5 is risky.** While it provides more training time per eval, Trial 2's late-training divergence went undetected longer. eval-every-3 is the safer choice.

### Suggested follow-ups

- **Extended timeout (600+ min):** Trial 1's metrics were still descending at timeout. A longer budget would likely close the volume gap further.
- **Gradient monitoring / early stopping:** Add per-epoch gradient norm logging even on non-eval epochs to detect divergence earlier.
- **Checkpoint on divergence detection:** Save model state when loss spikes above a threshold.
- **Adaptive eval frequency:** Eval every N epochs early, then switch to every-epoch eval in the final 100 epochs to capture terminal metrics precisely.

**Comment by tcapelle** [comment]
ADVISOR: Excellent finding — surface_mse=0.000266 beats baseline by 10.1%! However vol_mse=0.002193 exceeds the 0.002039 ceiling. The eval-every-3 throughput trick is clearly real.

**Follow-up:** Keep eval-every-3 but increase vol-weight to 12x (from 10x) to recover volume. Run:

```bash
cd target/icml2026 && python train.py --dataset airfrans --optimizer lion --lr 1.25e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --model-layers 4 --model-hidden-dim 256 --model-heads 4 --enable-fourier --epochs 999 --ema-decay 0.999 --use-lookahead --compile-model --vol-loss-weight 12 --eval-interval 3 --wandb-group af-eval-every-3-vol12x
```

**Target:** surface_mse < 0.000296 AND vol_mse < 0.002039 simultaneously.

**Comment by tcapelle** [comment]
ADVISOR: Excellent result — val surface_mse=0.000266 is a NEW AF BEST (-10.1% vs 0.000296 baseline). This is clearly a winner on the primary metric.

Before I merge, please run **full-eval TEST** on the Trial 1 best checkpoint (ep792, eval-every-3 run). We need the paper-facing test metrics:

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 --vol-loss-weight 10.0 \
  --load-checkpoint <path-to-best-checkpoint> --eval-only --max-eval-batches 0
```

If no checkpoint was saved, please re-run Trial 1 with `--save-checkpoint` and then full-eval.

Report: test_primary/surface_mse and test_primary/volume_mse from full eval. Submit for review when done.
